pennylane/data/base/attribute.py

```diff
@@ -0,0 +1,426 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains the base class for Dataset attribute types, and a class for
+attribute metadata."""
+
+import itertools
+import typing
+import warnings
+from abc import ABC, abstractmethod
+from collections.abc import MutableMapping
+from numbers import Number
+from types import MappingProxyType
+from typing import (
+    Any,
+    ClassVar,
+    Generic,
+    Iterator,
+    Literal,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from pennylane.data.base._zarr import zarr
+from pennylane.data.base.typing_util import (
+    UNSET,
+    T,
+    Zarr,
+    ZarrAny,
+    ZarrGroup,
+    get_type_str,
+    resolve_special_type,
+)
+
+
+class AttributeInfo(MutableMapping):
+    """Contains metadata that may be assigned to a dataset
+    attribute. Is stored in the Zarr object's ``attrs`` dict.
+
+    Attributes:
+        attrs_namespace: Keys for this class's attributes will
+            be prefixed with this string in ``attrs_bind``.
+        attrs_bind: The Zarr attrs dict that this instance is bound to,
+            or any mutable mapping
+        py_type: Type annotation for this attribute
+        doc: Documentation for this attribute
+        **kwargs: Extra metadata to include. Must be a string, number
+            or numpy array
+    """
+
+    attrs_namespace: ClassVar[str] = "qml.data."
+
+    attrs_bind: typing.MutableMapping[str, Any]
+
+    doc: Optional[str]
+
+    @overload
+    def __init__(
+        self,
+        attrs_bind: Optional[typing.MutableMapping[str, Any]] = None,
+        *,
+        doc: Optional[str] = None,
+        py_type: Optional[str] = None,
+        **kwargs: Any,
+    ):
+        ...
+
+    @overload
+    def __init__(self):
+        ...
```

Comment:
The first `overload` is just to get hints for the known keyword arguments, while the actual init keeps them in ``kwargs``. The second overload is just there because it's mandatory to have at least two overloads